packages/embedder/embedder.ts

```diff
@@ -0,0 +1,250 @@
+import tracer from 'dd-trace'
+import Redlock, {RedlockAbortSignal} from 'redlock'
+
+import 'parabol-server/initSentry'
+import getKysely from 'parabol-server/postgres/getKysely'
+import RedisInstance from 'parabol-server/utils/RedisInstance'
+import {DB} from 'parabol-server/postgres/pg'
+import getDataLoader from 'parabol-server/graphql/getDataLoader'
+import {refreshEmbeddingsMeta as refreshRetroDiscussionTopicsMeta} from './indexing/retrospectiveDiscussionTopic'
+import {orgIdsWithFeatureFlag} from './indexing/orgIdsWithFeatureFlag'
+import getModelManager, {ModelManager} from './ai_models/ModelManager'
+import {countWords} from './indexing/countWords'
+import {createEmbeddingTextFrom} from './indexing/createEmbeddingTextFrom'
+import {DataLoaderWorker} from 'parabol-server/graphql/graphql'
+import {
+  selectJobQueueItemById,
+  selectMetadataByJobQueueId,
+  updateJobState
+} from './indexing/embeddingsTablesOps'
+import {selectMetaToQueue} from './indexing/embeddingsTablesOps'
+import {insertNewJobs} from './indexing/embeddingsTablesOps'
+import {completeJobTxn} from './indexing/embeddingsTablesOps'
+
+/*
+ * TODO List
+ * - [ ] implement a clean-up function that re-queues items that haven't transitioned
+ *       to a completed state, or that failed
+ */
+
+export type DBInsert = {
+  [K in keyof DB]: Insertable<DB[K]>
+}
+
+const POLLING_PERIOD_SEC = 60 // How often do we try to grab the lock and re-index?
+const Q_MAX_LENGTH = 100 // How many EmbeddingIndex items do we batch in redis?
+const WORD_COUNT_TO_TOKEN_RATIO = 3.0 / 2 // We multiple the word count by this to estimate token count
+
+const {AI_EMBEDDER_ENABLED} = process.env
+const {SERVER_ID} = process.env
+
+tracer.init({
+  service: `embedder`,
+  appsec: process.env.DD_APPSEC_ENABLED === 'true',
+  plugins: false,
+  version: process.env.npm_package_version
+})
+tracer.use('pg')
+
+export const pg = getKysely()
+const dataLoader = getDataLoader() as DataLoaderWorker
+
+const redisClient = new RedisInstance(`embedder-${SERVER_ID}`)
+const redlock = new Redlock([redisClient], {
+  driftFactor: 0.01,
+  retryCount: 10,
+  retryDelay: 250,
+  retryJitter: 50,
+  automaticExtensionThreshold: 500
+})
+
+const refreshMetadata = async () => {
+  await refreshRetroDiscussionTopicsMeta(dataLoader)
+  // In the future, other sorts of objects to index could be added here...
+}
+const maybeQueueMetadataItems = async (modelManager: ModelManager) => {
+  const queueLength = await redisClient.zcard('embedder:queue')
+  if (queueLength >= Q_MAX_LENGTH) return
+  const itemCountToQueue = Q_MAX_LENGTH - queueLength
+  const configuredModels = [...modelManager.getEmbeddingModelsIter()].map((m) => m.getTableName())
+  const orgIds = await orgIdsWithFeatureFlag()
+
+  // For each configured embedding model, select rows from EmbeddingsMetadata
+  // that haven't been calculated nor exist in the EmbeddingsJobQueue yet
+  //
+  // Notes:
+  //   * `em.models @> ARRAY[v.model]` is an indexed query
```

Comment:
I've tested this on a few hundred thousand rows on my system and EXPLAINing the queries it always looked to do the right thing... but let's keep an eye on it

Reply:
this query runs a _little_ slower than just doing a join, but not enough to care. That said, if denormalizing our data isn't getting us a speed boost, why not get rid of `models`? It introduces a 2nd source of truth & makes queries more complex to write. Check out the queries used for benchmarking

CURRENT
```sql
select "em".*, "model" from "EmbeddingsMetadata" as "em" 
left join lateral unnest(ARRAY[['Embeddings_ember_1']]::varchar[]) as "model" on true  
where ((not "em"."models" <@ ARRAY["model"]::varchar[] or "em"."models" is null)
and not exists (select "ejq"."id" from "EmbeddingsJobQueue" as "ejq" 
 where "em"."objectType" = "ejq"."objectType" 
 and "em"."refId" = "ejq"."refId" and "ejq"."model" = "model"))
```

PROPOSED
```sql
select * from "EmbeddingsMetadata" as em
left join "Embeddings_ember_1" as ember on ember."embeddingsMetadataId" = em.id
left join "EmbeddingsJobQueue" as ejq on ejq."embeddingsMetadataId" = em.id
where ember.id is null and ejq.id is null
```

Reply:
I have a tendency to agree with you here. I suppose since we're queuing jobs one model at a time anyway, it's effort equivalent to just:

```
   FOR MODEL_TABLE_NAME IN MODEL_TABLE_NAMES:
      select * from "EmbeddingsMetadata" as em
      left join MODEL_TABLE_NAME as mt on mt."embeddingsMetadataId" = em.id
      left join "EmbeddingsJobQueue" as ejq on ejq."embeddingsMetadataId" = em.id
      where mt.id is null and ejq.id is null
```

Let's do this as a separate PR just to keep things need. I'll open the issue.

Reply:
Linking this here: https://github.com/ParabolInc/parabol/issues/9449